NSSServices/Controllers/VariablesController.cs

```diff
@@ -64,14 +69,189 @@ public override async Task<IActionResult> Get(int id)
+                for (int i = 0; i < newVarTypeToGrabIDList.Count(); i++)
+                {
+                    newVariableList[i].VariableTypeID = newVarTypeToGrabIDList[i].ID;
+
+                    returnVariableWithUnitList[i].ID = newVarTypeToGrabIDList[i].ID;
+                }
+
+                agent.Add(newVariableList).Wait();
+
+                return Ok(returnVariableWithUnitList);
+            }
+            catch (Exception ex)
+            {
+                return await HandleExceptionAsync(ex);
+            }
+        }
+
+        [HttpPut("{id}", Name = "Edit Variable")]
+        [Authorize(Policy = Policy.AdminOnly)]
+        [APIDescription(type = DescriptionType.e_link, Description = "/Docs/Variables/Edit.md")]
+        public async Task<IActionResult> Put(int id, [FromBody] VariableWithUnit entity)
+        {
+            try
+            {
+                if (id < 0 || !isValid(entity)) return new BadRequestResult(); // This returns HTTP 404
+
+                Variable newVariable = new Variable
+                {
+                    VariableTypeID = id,
+                    UnitTypeID = entity.UnitTypeID,
+                    Comments = "Default unit"
+                };
+
+                VariableType newVariableType = new VariableType
+                {
+                    ID = id,
+                    Name = entity.Name,
+                    Code = entity.Code,
+                    Description = entity.Description
+                };
+
+                var newVarTypeToGrabID = await shared.Update(id, newVariableType);
+                var varID = agent.GetVariable(id);
+                await agent.Update(varID.ID, newVariable);
+
+                VariableWithUnit returnVariableWithUnit = new VariableWithUnit
+                {
+                    ID = newVarTypeToGrabID.ID,
+                    Name = newVariableType.Name,
+                    Code = newVariableType.Code,
+                    Description = newVariableType.Description,
+                    UnitTypeID = newVariable.UnitTypeID
+                };
+
+                return Ok(returnVariableWithUnit);
+            }
+            catch (Exception ex)
+            {
+                return await HandleExceptionAsync(ex);
+            }
+
+        }
+
+        [HttpDelete("{id}", Name = "Delete Variable")]
+        [Authorize(Policy = Policy.AdminOnly)]
+        [APIDescription(type = DescriptionType.e_link, Description = "/Docs/Variables/Delete.md")]
+        public async Task<IActionResult> Delete(int id)
+        {
+            try
+            {
+                if (id < 1) return new BadRequestResult();
+
+                var returnVar = agent.DeleteVariable(id);
+                if (returnVar != null)
+                {
+                    shared.DeleteVariableType(id).Wait();
+                    return Ok("Deleted ID: " + id);
+                }
+                return Ok("Variable and VariableType not deleted because of Foreign Key Constraint");
```

Comment:
I did the variable deletion first because the deletion will fail and return a null which then means the variable type will not be deleted.

I was doing the opposite at first, deleting the type then the variable but it was leaving variables not deleted since the variable had dependencies. The Task DeleteVariable code in NSSAgent is checking for any dependencies before deleting a Variable so if there are no dependencies, the Variable Type should be safe to delete.